.github/workflows/release.yaml

```diff
@@ -0,0 +1,79 @@
+name: Release
+on:
+  workflow_dispatch:
+    inputs:
+      releaseVersion:
+        description: 'Release Version'
+        required: true
+        default: '0.0.1'
+      snapshotVersion:
+        description: 'Snapshot Version'
+        required: true
+        default: '0.1.0-SNAPSHOT'
+      branch:
+        description: 'Branch to Release'
+        required: false
+        default: 'main'
+permissions:
+  contents: write
+jobs:
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.inputs.branch }}
+      - name: Set up Java
+        uses: actions/setup-java@v4
+        with:
+          java-version: '17'
+          distribution: 'temurin'
+          cache: maven
+      - name: Update Sources
+        run: |
+          .github/scripts/setVersions.sh "${{ github.event.inputs.releaseVersion }}" "v${{ github.event.inputs.releaseVersion }}" "v${{ github.event.inputs.releaseVersion }}"
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v3
+      - name: Test Build Image
+        uses: docker/build-push-action@v6
+        with:
+          context: data-generator/
+          platforms: linux/amd64,linux/arm64
+          push: false
+          file: data-generator/Dockerfile
+      - name: Push Tag
+        run: |
+          git push
+          git push origin "v${{ github.event.inputs.releaseVersion }}"
+      - name: Image metadata
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: |
+            ${{ secrets.IMAGE_REPO_HOSTNAME }}/${{ secrets.IMAGE_REPO_NAMESPACE }}/flink-examples-data-generator
+          tags: |
+            type=raw,value=v${{ github.event.inputs.releaseVersion }}
+            type=raw,value=latest
+      - name: Login to Quay
+        uses: docker/login-action@v3
+        with:
+          registry: "${{ secrets.IMAGE_REPO_HOSTNAME }}"
+          username: "${{ secrets.IMAGE_REPO_USERNAME }}"
+          password: "${{ secrets.IMAGE_REPO_PASSWORD }}"
+      - name: Build and Push Image
+        uses: docker/build-push-action@v6
+        with:
+          context: data-generator/
+          platforms: linux/amd64,linux/arm64
+          push: true
+          file: data-generator/Dockerfile
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+      - name: Update Sources for Development
+        run: |
+          .github/scripts/setVersions.sh "${{ github.event.inputs.snapshotVersion }}" "latest"
```

Comment:
Why should we revert to the latest/snapshot versions? I think in this release branch, we should keep using the image from this version. Supposedly we're in v0.1.0 branch, and after releasing, this v0.1.0 branch will use the latest one (supposedly it's v0.3.0)), which doesn't make sense to me.

Reply:
hmm I hadn't really considered release branches, I was thinking only of the main workflow. As we are accepting PRs into main it would be nice if users can execute the README as it is in main and have it execute the latest code from main. Maybe it's better done by tagging the image with the branch name rather than latest. So you could have a flow like:

1. run an action to create a release PR, updating the versions to X.Y.Z.
2. merge the PR
3. integration.yaml runs CI and pushes `main` to quay.io
4. releaser creates the release tag vX.Y.Z and pushes it to github
5. integration.yaml runs CI against the new tag and pushes vX.Y.Z to quay.io
6. releaser runs an action to create development PR, updating to version A.B.C-SNAPSHOT and image references to `main`
7. merge the PR
8. integration.yaml runs CI and pushes `main` to quay.io

I think these actions could also be applied to a release branch. The `create development PR` action would be told the target branch and so it can update the image reference appropriately.

Reply:
Yes, sounds good to me. Also, we should have a README somewhere to guide people how to run the release.

Reply:
Yep, have added a README in the replacement PR #34